diff --git a/config.sub b/config.sub
index ba37cf9..52a9ec6 100755
--- a/config.sub
+++ b/config.sub
@@ -1042,7 +1042,7 @@ case $basic_machine in
 		;;
 	ppc64)	basic_machine=powerpc64-unknown
 		;;
-	ppc64-*) basic_machine=powerpc64-`echo "$basic_machine" | sed 's/^[^-]*-//'`
+	ppc64-* | ppc64p7-*) basic_machine=powerpc64-`echo "$basic_machine" | sed 's/^[^-]*-//'`
 		;;
 	ppc64le | powerpc64little)
 		basic_machine=powerpc64le-unknown
diff --git a/configure.ac b/configure.ac
index 2535969..9fb3193 100644
--- a/configure.ac
+++ b/configure.ac
@@ -765,9 +765,9 @@ cat >> conftest.c <<EOF
         alpha-linux-gnu
 # elif defined(__ARM_EABI__) && defined(__ARM_PCS_VFP)
 #  if defined(__ARMEL__)
-        arm-linux-gnueabihf
+        arm-linux-gnueabi
 #  else
-        armeb-linux-gnueabihf
+        armeb-linux-gnueabi
 #  endif
 # elif defined(__ARM_EABI__) && !defined(__ARM_PCS_VFP)
 #  if defined(__ARMEL__)
@@ -807,7 +807,7 @@ cat >> conftest.c <<EOF
 #  elif _MIPS_SIM == _ABIN32
         mips64el-linux-gnuabin32
 #  elif _MIPS_SIM == _ABI64
-        mips64el-linux-gnuabi64
+        mips64el-linux-gnu
 #  else
 #   error unknown platform triplet
 #  endif
@@ -817,7 +817,7 @@ cat >> conftest.c <<EOF
 #  elif _MIPS_SIM == _ABIN32
         mips64-linux-gnuabin32
 #  elif _MIPS_SIM == _ABI64
-        mips64-linux-gnuabi64
+        mips64-linux-gnu
 #  else
 #   error unknown platform triplet
 #  endif
@@ -827,9 +827,9 @@ cat >> conftest.c <<EOF
         powerpc-linux-gnuspe
 # elif defined(__powerpc64__)
 #  if defined(__LITTLE_ENDIAN__)
-        powerpc64le-linux-gnu
+        ppc64le-linux-gnu
 #  else
-        powerpc64-linux-gnu
+        ppc64-linux-gnu
 #  endif
 # elif defined(__powerpc__)
         powerpc-linux-gnu
