From e03f3216bbf99b64cfa27343d20534fb306ea2e5 Mon Sep 17 00:00:00 2001
From: Dino Viehland <dinoviehland@gmail.com>
Date: Mon, 27 Jan 2020 14:04:56 -0800
Subject: [PATCH] 00340: Add test.test_import.data.unwritable package to
 makefile

---
 Makefile.pre.in | 1 +
 1 file changed, 1 insertion(+)

diff --git a/Makefile.pre.in b/Makefile.pre.in
index af24a2fe2a..aa20b65aac 100644
--- a/Makefile.pre.in
+++ b/Makefile.pre.in
@@ -1330,6 +1330,7 @@ LIBSUBDIRS=	tkinter tkinter/test tkinter/test/test_tkinter \
 		test/test_import/data/circular_imports/subpkg \
 		test/test_import/data/package \
 		test/test_import/data/package2 \
+		test/test_import/data/unwritable \
 		importlib \
 		importlib/metadata \
 		test/test_importlib \
-- 
2.24.1

